File: universal/Decoder.cpp
#include "pch.h"
#include "Decoder.h"
#include "DecoderSettings.h"
#include "Utilities.h"
#include <windows.storage.streams.h>

using namespace ManagedLzma::Universal;
using namespace Platform;

#pragma region Decoder

Decoder::Decoder(DecoderSettings^ settings)
	: mSettings(settings)
{
	if (!settings)
		throw ref new InvalidArgumentException();

	LzmaDec_Construct(&mDecoder);
	byte props[LZMA_PROPS_SIZE];
	settings->WriteTo(props);
	auto res = LzmaDec_Allocate(&mDecoder, props, LZMA_PROPS_SIZE, kAllocImpl);
	if (res != SZ_OK)
		throw CreateException(res);
}

Decoder::~Decoder()
{
	for (;;)
	{
		concurrency::critical_section::scoped_lock lock(mSync);

		if (!mRunning)
			break;

		// TODO: stop running decoder and wait until it completes
		RoFailFastWithErrorContext(E_NOTIMPL);
	}

	LzmaDec_Free(&mDecoder, kAllocImpl);
}

Windows::Foundation::IAsyncAction^ Decoder::CancelAsync()
{
	// TODO: we should cancel all outstanding tasks and then complete this action
	throw ref new NotImplementedException();
}

Windows::Foundation::IAsyncAction^ Decoder::CompleteInputAsync()
{
	concurrency::critical_section::scoped_lock lock(mSync);

	mFlushed = true;

	return concurrency::create_async([this] {
		RoFailFastWithErrorContext(E_NOTIMPL);
	});
}

Windows::Foundation::IAsyncOperationWithProgress<uint32_t, uint32_t>^ Decoder::ProvideInputAsync(Windows::Storage::Streams::IBuffer^ buffer)
{
	if (!buffer)
		throw ref new NullReferenceException(L"buffer cannot be null");

	return concurrency::create_async([&](concurrency::progress_reporter<uint32_t> progress) {
		auto pInputFrame = std::make_shared<InputFrame>();
		pInputFrame->mBuffer = buffer;
		pInputFrame->mOffset = 0;
		pInputFrame->mEnding = buffer->Length;
		pInputFrame->mProgress = progress;
		PushInputFrame(pInputFrame);
		return concurrency::create_task(pInputFrame->mCompletion);
	});
}

Windows::Foundation::IAsyncOperationWithProgress<Windows::Storage::Streams::IBuffer^, uint32_t>^ Decoder::ReadOutputAsync(
	Windows::Storage::Streams::IBuffer^ buffer, uint32_t length, Windows::Storage::Streams::InputStreamOptions options)
{
	if (!buffer)
		throw ref new NullReferenceException(L"buffer cannot be null");

	if (length > buffer->Capacity - buffer->Length)
		throw ref new InvalidArgumentException(L"buffer not large enough");

	return concurrency::create_async([&](concurrency::progress_reporter<uint32_t> progress) {
		auto pOutputFrame = std::make_shared<OutputFrame>();
		pOutputFrame->mBuffer = buffer;
		pOutputFrame->mOffset = buffer->Length;
		pOutputFrame->mEnding = pOutputFrame->mOffset + length;
		pOutputFrame->mProgress = progress;
		pOutputFrame->mOptions = options;
		PushOutputFrame(pOutputFrame);
		return concurrency::create_task(pOutputFrame->mCompletion);
	});
}

void Decoder::PushInputFrame(const std::shared_ptr<InputFrame>& frame)
{
	concurrency::critical_section::scoped_lock lock(mSync);
	mInputQueue.push(frame);
	TryStartOperation();
}

void Decoder::PushOutputFrame(const std::shared_ptr<OutputFrame>& frame)
{
	concurrency::critical_section::scoped_lock lock(mSync);
	mOutputQueue.push(frame);
	TryStartOperation();
}

void Decoder::TryStartOperation()
{
	// Caller locks.
	//concurrency::critical_section::scoped_lock lock(mSync);

	if (!mRunning)
	{
		if (!mStarted)
		{
			mStarted = true;
			LzmaDec_Init(&mDecoder);
		}

		mOperation = concurrency::create_task([this] {
			do { WriteOutput(); } while (DecodeInput());
		});
	}
}

void Decoder::WriteOutput()
{
	while (mDecoderPosition != mDecoder.dicPos)
	{
		OutputFrame* pOutputFrame;
		{
			// Access to the output queue must be locked.
			concurrency::critical_section::scoped_lock lock(mSync);

			if (!mRunning || !mStarted)
				RoFailFastWithErrorContext(E_UNEXPECTED);

			if (mOutputQueue.empty())
				break;

			pOutputFrame = mOutputQueue.front().get();
		}

		HRESULT hr = S_OK;
		Microsoft::WRL::ComPtr<Windows::Storage::Streams::IBufferByteAccess> pOutputBufferAccess;
		if (FAILED(hr = reinterpret_cast<IUnknown*>(pOutputFrame->mBuffer)->QueryInterface(IID_PPV_ARGS(&pOutputBufferAccess))))
			throw Exception::CreateException(hr);

		byte* pOutputBuffer = NULL;
		if (FAILED(hr = pOutputBufferAccess->Buffer(&pOutputBuffer)))
			throw Exception::CreateException(hr);

		auto capacity = pOutputFrame->mEnding - pOutputFrame->mOffset;
		auto copySize = std::min(capacity, mDecoder.dicPos - mDecoderPosition);
		memcpy(pOutputBuffer + pOutputFrame->mOffset, mDecoder.dic + mDecoderPosition, copySize);

		pOutputBufferAccess.Reset();
		pOutputFrame->mOffset += copySize;
		mDecoderPosition += copySize;

		if (copySize == capacity || (pOutputFrame->mOptions & Windows::Storage::Streams::InputStreamOptions::Partial) == Windows::Storage::Streams::InputStreamOptions::Partial)
		{
			pOutputFrame->mCompletion.set(pOutputFrame->mBuffer);

			// Access to the output queue must be locked.
			concurrency::critical_section::scoped_lock lock(mSync);

			if (!mRunning || !mStarted)
				RoFailFastWithErrorContext(E_UNEXPECTED);

			mOutputQueue.pop();
		}
		else
		{
			auto progress = (100.0f * pOutputFrame->mOffset) / pOutputFrame->mEnding;
			pOutputFrame->mProgress.report((uint32_t)lround(progress));
		}
	}
}

bool Decoder::DecodeInput()
{
	ELzmaFinishMode mode;
	InputFrame* pInputFrame;
	{
		// Access to the input queue must be locked.
		concurrency::critical_section::scoped_lock lock(mSync);

		if (!mRunning || !mStarted)
			RoFailFastWithErrorContext(E_UNEXPECTED);

		if (mInputQueue.empty())
		{
			mRunning = false;
			return false;
		}

		mode = mFlushed && mInputQueue.size() == 1 ? LZMA_FINISH_END : LZMA_FINISH_ANY;
		pInputFrame = mInputQueue.front().get();
	}

	HRESULT hr = S_OK;
	Microsoft::WRL::ComPtr<Windows::Storage::Streams::IBufferByteAccess> pInputBufferAccess;
	if (FAILED(hr = reinterpret_cast<IUnknown*>(pInputFrame->mBuffer)->QueryInterface(IID_PPV_ARGS(&pInputBufferAccess))))
		throw Exception::CreateException(hr);

	byte* pInputBuffer = NULL;
	if (FAILED(hr = pInputBufferAccess->Buffer(&pInputBuffer)))
		throw Exception::CreateException(hr);

	if (mDecoder.dicPos == mDecoder.dicBufSize) mDecoder.dicPos = 0;
	size_t dicLimit = std::min(mDecoderPosition + mTotalOutputCapacity, mDecoder.dicBufSize);
	size_t inputSize = pInputFrame->mEnding - pInputFrame->mOffset;
	ELzmaStatus status = LZMA_STATUS_NOT_SPECIFIED;

	SRes res = LzmaDec_DecodeToDic(&mDecoder, 0, pInputBuffer + pInputFrame->mOffset, &inputSize, LZMA_FINISH_ANY, &status);
	if (res != SZ_OK)
	{
		if (res != SZ_ERROR_DATA)
			RoFailFastWithErrorContext(E_UNEXPECTED);

		throw Exception::CreateException(HRESULT_FROM_WIN32(ERROR_INVALID_DATA));
	}

	switch (status)
	{
	case LZMA_STATUS_FINISHED_WITH_MARK:
	case LZMA_STATUS_NOT_FINISHED:
	case LZMA_STATUS_NEEDS_MORE_INPUT:
	case LZMA_STATUS_MAYBE_FINISHED_WITHOUT_MARK:
		mDecoderState = status;
		break;

	default:
		RoFailFastWithErrorContext(E_UNEXPECTED);
	}

	pInputBufferAccess.Reset(); // we are done with the buffer for now
	pInputFrame->mOffset += inputSize;

	if (pInputFrame->mOffset == pInputFrame->mEnding)
	{
		pInputFrame->mCompletion.set(pInputFrame->mEnding);

		// Access to the input queue must be locked.
		concurrency::critical_section::scoped_lock lock(mSync);

		if (!mRunning || !mStarted)
			RoFailFastWithErrorContext(E_UNEXPECTED);

		mInputQueue.pop();
	}
	else
	{
		auto progress = (100.0f * pInputFrame->mOffset) / pInputFrame->mEnding;
		pInputFrame->mProgress.report((uint32_t)lround(progress));
	}

	return true;
}

#pragma endregion

#pragma region DecoderInputStream

DecoderInputStream::DecoderInputStream(Decoder^ decoder)
	: mDecoder(decoder)
{
}

DecoderInputStream::~DecoderInputStream()
{
	mDecoder = nullptr;
}

Windows::Foundation::IAsyncOperation<bool>^ DecoderInputStream::FlushAsync()
{
	return concurrency::create_async([this] { return concurrency::create_task(mDecoder->CompleteInputAsync()).then([] { return true; }); });
}

Windows::Foundation::IAsyncOperationWithProgress<uint32_t, uint32_t>^ DecoderInputStream::WriteAsync(Windows::Storage::Streams::IBuffer^ buffer)
{
	return mDecoder->ProvideInputAsync(buffer);
}

#pragma endregion

#pragma region DecoderInputStream

DecoderOutputStream::DecoderOutputStream(Decoder^ decoder)
	: mDecoder(decoder)
{
}

DecoderOutputStream::~DecoderOutputStream()
{
	mDecoder = nullptr;
}

Windows::Foundation::IAsyncOperationWithProgress<Windows::Storage::Streams::IBuffer^, uint32_t>^ DecoderOutputStream::ReadAsync(
	Windows::Storage::Streams::IBuffer^ buffer, uint32_t length, Windows::Storage::Streams::InputStreamOptions options)
{
	return mDecoder->ReadOutputAsync(buffer, length, options);
}

#pragma endregion

File: master/api/lzma/Decoder.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace ManagedLzma.LZMA
{
    public sealed class Decoder : IDisposable
    {
        private sealed class InputFrame
        {
            public byte[] mBuffer;
            public int mOrigin;
            public int mOffset;
            public int mEnding;
            public TaskCompletionSource<object> mCompletion = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
        }

        private sealed class OutputFrame
        {
            public byte[] mBuffer;
            public int mOrigin;
            public int mOffset;
            public int mEnding;
            public TaskCompletionSource<int> mCompletion = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
            public ReadMode mMode;
        }

        // immutable
        private readonly object mSyncObject;
        private readonly DecoderSettings mSettings;
        private readonly Action mDecodeAction;

        // multithreaded access (under lock)
        private Task mDecodeTask;
        private Queue<InputFrame> mInputQueue;
        private Queue<OutputFrame> mOutputQueue;
        private int mTotalOutputCapacity;
        private bool mStarted;
        private bool mRunning;
        private bool mFlushed;

        // owned by decoder task
        private Master.LZMA.CLzmaDec mDecoder;
        private Master.LZMA.ELzmaStatus mStatus;
        private int mDecoderPosition;

        public Decoder(DecoderSettings settings)
        {
            if (settings == null)
                throw new ArgumentNullException(nameof(settings));

            mSettings = settings;
            mSyncObject = new object();
            mDecodeAction = new Action(Decode);
            mDecoder = new Master.LZMA.CLzmaDec();
            mDecoder.LzmaDec_Construct();
            if (mDecoder.LzmaDec_Allocate(settings.ToArray(), Master.LZMA.LZMA_PROPS_SIZE, Master.LZMA.ISzAlloc.SmallAlloc) != Master.LZMA.SZ_OK)
                throw new InvalidOperationException();
        }

        public void Dispose()
        {
            for (;;)
            {
                lock (mSyncObject)
                {
                    if (!mRunning)
                        break;


                    throw new NotImplementedException();
                }
            }

            mDecoder.LzmaDec_Free(Master.LZMA.ISzAlloc.SmallAlloc);
        }

        public Task DisposeAsync()
        {
            throw new NotImplementedException();
        }

        public Task CompleteInputAsync()
        {
            throw new NotImplementedException();
        }

        public Task WriteInputAsync(byte[] buffer, int offset, int length)
        {
            if (buffer == null)
                throw new ArgumentNullException(nameof(buffer));

            if (offset < 0 || offset > buffer.Length)
                throw new ArgumentOutOfRangeException(nameof(offset));

            if (length < 0 || length > buffer.Length - offset)
                throw new ArgumentOutOfRangeException(nameof(length));

            var frame = new InputFrame();
            frame.mBuffer = buffer;
            frame.mOrigin = offset;
            frame.mOffset = offset;
            frame.mEnding = offset + length;
            PushInputFrame(frame);
            return frame.mCompletion.Task;
        }

        public Task<int> ReadOutputAsync(byte[] buffer, int offset, int length, ReadMode mode)
        {
            if (buffer == null)
                throw new ArgumentNullException(nameof(buffer));

            if (offset < 0 || offset > buffer.Length)
                throw new ArgumentOutOfRangeException(nameof(offset));

            if (length < 0 || length > buffer.Length - offset)
                throw new ArgumentOutOfRangeException(nameof(length));

            if (mode != ReadMode.FillBuffer && mode != ReadMode.ReturnEarly)
                throw new ArgumentOutOfRangeException(nameof(mode));

            var frame = new OutputFrame();
            frame.mBuffer = buffer;
            frame.mOrigin = offset;
            frame.mOffset = offset;
            frame.mEnding = offset + length;
            frame.mMode = mode;
            PushOutputFrame(frame);
            return frame.mCompletion.Task;
        }

        private void PushInputFrame(InputFrame frame)
        {
            lock (mSyncObject)
            {
                mInputQueue.Enqueue(frame);
                TryStartDecoding();
            }
        }

        private void PushOutputFrame(OutputFrame frame)
        {
            lock (mSyncObject)
            {
                mTotalOutputCapacity += frame.mEnding - frame.mOffset;
                mOutputQueue.Enqueue(frame);
                TryStartDecoding();
            }
        }

        private void TryStartDecoding()
        {
            System.Diagnostics.Debug.Assert(Monitor.IsEntered(mSyncObject));

            if (!mRunning)
            {
                mRunning = true;

                if (!mStarted)
                {
                    mStarted = true;
                    mDecoder.LzmaDec_Init();
                }

                System.Diagnostics.Debug.Assert(mDecodeTask == null);
                mDecodeTask = Task.Run(mDecodeAction);
            }
        }

        private void Decode()
        {
            do { WriteOutput(); }
            while (DecodeInput());
        }

        private void WriteOutput()
        {
            while (mDecoderPosition != mDecoder.mDicPos)
            {
                OutputFrame frame;
                lock (mSyncObject)
                {
                    System.Diagnostics.Debug.Assert(mStarted && mRunning);

                    if (mOutputQueue.Count == 0)
                        break;

                    frame = mOutputQueue.Peek();
                }

                var capacity = frame.mEnding - frame.mOffset;
                var copySize = Math.Min(capacity, checked((int)mDecoder.mDicPos) - mDecoderPosition);
                Buffer.BlockCopy(mDecoder.mDic.mBuffer, mDecoder.mDic.mOffset + mDecoderPosition, frame.mBuffer, frame.mOffset, copySize);

                frame.mOffset += copySize;
                mDecoderPosition += copySize;

                if (copySize == capacity || frame.mMode == ReadMode.ReturnEarly)
                {
                    frame.mCompletion.SetResult(frame.mOffset - frame.mOrigin);

                    lock (mSyncObject)
                    {
                        System.Diagnostics.Debug.Assert(mStarted && mRunning);
                        var other = mOutputQueue.Dequeue();
                        System.Diagnostics.Debug.Assert(other == frame);
                    }
                }
            }
        }

        private bool DecodeInput()
        {
            var mode = Master.LZMA.ELzmaFinishMode.LZMA_FINISH_ANY;
            var capacity = default(int);
            var frame = default(InputFrame);

            lock (mSyncObject)
            {
                System.Diagnostics.Debug.Assert(mStarted && mRunning);

                if (mInputQueue.Count == 0)
                {
                    mRunning = false;
                    return false;
                }

                if (mFlushed && mInputQueue.Count == 1)
                    mode = Master.LZMA.ELzmaFinishMode.LZMA_FINISH_END;

                capacity = mTotalOutputCapacity;
                frame = mInputQueue.Peek();
            }

            if (mDecoder.mDicPos == mDecoder.mDicBufSize)
                mDecoder.mDicPos = 0;

            long limit = Math.Min(mDecoderPosition + capacity, mDecoder.mDicBufSize);
            long input = frame.mEnding - frame.mOffset;
            var res = mDecoder.LzmaDec_DecodeToDic(limit, P.From(frame.mBuffer, frame.mOffset), ref input, mode, out mStatus);
            if (res != Master.LZMA.SZ_OK)
            {
                System.Diagnostics.Debug.Assert(res == Master.LZMA.SZ_ERROR_DATA);
                throw new InvalidDataException();
            }

            frame.mOffset += checked((int)input);

            if (frame.mOffset == frame.mEnding)
            {
                frame.mCompletion.SetResult(null);

                lock (mSyncObject)
                {
                    System.Diagnostics.Debug.Assert(mStarted && mRunning);
                    var other = mInputQueue.Dequeue();
                    System.Diagnostics.Debug.Assert(other == frame);
                }
            }

            return true;
        }
    }
}
